invenio_app_ils/anonymization.py

```diff
@@ -0,0 +1,193 @@
+    if patron_pid is None:
+        raise ValueError("No patron pid was provided.")
+
+    patron = get_patron_or_unknown(patron_pid)
+    if not patron and not force:
+        return None
+
+    patron_object = None
+    try:
+        patron_object = current_app_ils.patron_cls.get_patron(patron_pid)
+    except PatronNotFoundError:
+        if not force:
+            raise PatronNotFoundError(patron_pid)
+
+    if (
+        get_loans_by_patron_pid(patron_pid)
+        .filter("term", state="ITEM_ON_LOAN")
+        .count()
+    ):
+        raise AssertionError(
+            "Cannot delete user %s: they have ongoing loans." % (patron_pid)
+        )
+
+    # Serialize empty patron values
+    anonymous_patron_fields = get_anonymous_patron_dict(patron_pid)
+
+    patron_loans = get_loans_by_patron_pid(patron_pid).execute()
+
+    indices = 0
+
+    for hit in patron_loans:
+        loan = Loan.get_record_by_pid(hit.pid)
+        if loan["state"] == "PENDING":
+            params = deepcopy(loan)
+            params.update(
+                dict(
+                    cancel_reason="Loan cancelled to anonymize user.",
+                    transaction_user_pid=str(SystemAgent.id),
+                )
+            )
+            current_circulation.circulation.trigger(
+                loan, **dict(params, trigger="cancel")
+            )
+        loan["patron"] = anonymous_patron_fields
+        loan.commit()
+        current_circulation.loan_indexer().index(loan)
+        indices += 1
+
+    patron_borrowing_requests = (
+        BorrowingRequestsSearch().search_by_patron_pid(patron_pid).scan()
+    )
+
+    for hit in patron_borrowing_requests:
+        borrowing_request = BorrowingRequest.get_record_by_pid(hit.pid)
+        borrowing_request["patron"] = anonymous_patron_fields
+        borrowing_request.commit()
+        current_ils_ill.borrowing_request_indexer_cls().index(
+            borrowing_request
+        )
+        indices += 1
+
+    patron_document_requests = (
+        DocumentRequestSearch().search_by_patron_pid(patron_pid).scan()
+    )
+
+    for hit in patron_document_requests:
+        document_request = DocumentRequest.get_record_by_pid(hit.pid)
+        if document_request["state"] == "PENDING":
+            document_request["state"] = "REJECTED"
+            document_request["reject_reason"] = "USER_CANCEL"
+        document_request["patron"] = anonymous_patron_fields
+        document_request.commit()
+        current_app_ils.document_request_indexer.index(document_request)
+        indices += 1
+
+    patron_acquisitions = OrderSearch().search_by_patron_pid(patron_pid).scan()
+
+    for hit in patron_acquisitions:
+        acquisition = Order.get_record_by_pid(hit.pid)
+        acquisition.commit()
+        current_ils_acq.order_indexer.index(acquisition)
+        indices += 1
+
+    # Delete rows from db
+    dropped = 0
+
+    with db.session.begin_nested():
```

Comment:
IMO this part should be implemented in separate function, f.e. `delete_user_account`, because it is different than `anonymize_patron_data`, so possibly they could be used separately if for some reason it is needed. In general it would be a separation of concerns

Reply:
Yes, however it is unlikely that we would need to delete the user rows while keeping the index out-of-date. The only use case I can identify is if we would like to anonymize their visible personal data while preserving their profile.
@KonstantinaStoikou what do you think?

Reply:
I think it's better to have separate functions